fs: nvs: Add CRC to protect the data part of the NVS items

RICCIARDI-Adrien opened the pull request:
Add a CRC-32 to protect the data part of a NVS item.
This CRC is optional and can be enabled with the `CONFIG_NVS_DATA_CRC` configuration item.

Also added a test scenario and updated the NVS documentation to reflect the existence of this CRC.

RICCIARDI-Adrien committed:
fs: nvs: Add CRC-32 to protect data

Allow to protect the data part of each NVS item with a 32-bit CRC.
This uses 4 more bytes per NVS item.

Signed-off-by: Adrien Ricciardi <aricciardi@baylibre.com>

RICCIARDI-Adrien committed:
tests: fs: nvs: Add a test scenario for data CRC

Adapt the test_nvs_gc_corrupt_close_ate test to forge the data CRC.

Signed-off-by: Adrien Ricciardi <aricciardi@baylibre.com>

RICCIARDI-Adrien committed:
doc: storage: nvs: Tell about optional data CRC

The CONFIG_NVS_DATA_CRC configuration item allows to enable a CRC-32
on the data part of the NVS items.

Signed-off-by: Adrien Ricciardi <aricciardi@baylibre.com>

Laczen commented:
Hi @RICCIARDI-Adrien, thank you for the PR. The proposal is OK, but we should warn the user about the space consequences of adding a CRC32: the extra crc will take an extra write-block-size and this can be quite large.

Laczen commented on subsys/fs/nvs/Kconfig:
```diff
@@ -29,6 +29,14 @@ config NVS_LOOKUP_CACHE_SIZE
+	  Enable a CRC-32 on the data part of each NVS element.
+	  The ATE CRC is not impacted by this feature and stays the same.
+	  The CRC-32 is transparently stored at the end of the data field,
+	  in the NVS data section, so 4 more bytes are needed per NVS element.
```
not 4 more bytes, but an extra write-block is needed

RICCIARDI-Adrien replied:
The idea is to write the CRC just after the data, taking alignment into account. I'm closing this conversation because we found a way to achieve this.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -175,12 +175,25 @@ static int nvs_flash_ate_wrt(struct nvs_fs *fs, const struct nvs_ate *entry)
+
+#ifdef CONFIG_NVS_DATA_CRC
+	/* Do not append the CRC if the data is being deleted (i.e. len == 0) */
+	if (len > 0 && compute_crc) {
```
add brackets

RICCIARDI-Adrien replied:
Done

Laczen commented:
@RICCIARDI-Adrien, I think there is an alignment problem. The writing of data first and then the crc32 makes that you need separate alignments and adapted length calculation. This needs to be solved, so I'm marking it as requires changes for the moment.

Laczen reviewed (changes requested):
Alignment of data and crc needs to be addressed

RICCIARDI-Adrien commented:
Thank you for the quick review. The alignment problem is a good catch, I'm working on it !

RICCIARDI-Adrien commented:
I reworked the `nvs_flash_al_wrt()` function to be able to append a CRC when needed. This way, only 4 more bytes (with extra padding if needed) are needed instead of a new `write-block` + padding (if needed).
I also added more tests to make everything is working fine.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -1198,14 +1250,38 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, uint16_t id, void *data, size_t len,
+		goto err;
+	}
+
+	computed_data_crc = crc32_ieee(data, wlk_ate.len - NVS_DATA_CRC_SIZE);
```
This calculation will fail if `len < wlk_ate.len`. The data buffer size can be smaller then `ate.len`.

RICCIARDI-Adrien replied:
You're correct ! Are you OK to disable CRC computation for partial reads ? CRC computation will be checked only for complete reads, which makes sense because if the CRC was implemented in a Zephyr application, the application would have need to read entirely the data to retrieve the CRC stored at the end.
I can update the documentation accordingly.

Laczen replied:
> You're correct ! Are you OK to disable CRC computation for partial reads ? CRC computation will be checked only for complete reads, which makes sense because if the CRC was implemented in a Zephyr application, the application would have need to read entirely the data to retrieve the CRC stored at the end. I can update the documentation accordingly.

Yes, I am OK with that.

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -128,13 +135,32 @@ static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
+		/* Write the last data part (if any), followed by the CRC (if any), followed by
+		 * the padding needed to respect the write block size (if any)
+		 */
+		len += unaligned_bytes_count;
```
I think this computation is wrong and should be :
len += fs->flash_parameters->write_block_size - unaligned_bytes_count;
Example :
write_block_size = 8
len = 7
len += sizeof(data_crc); => len = 11
unaligned_bytes_count = 3
len + unaligned_bytes_count; => len = 14

writes should be always a multiple of write_block_size
len + write_block_size - unaligned_bytes_count = 16

RICCIARDI-Adrien replied:
Correct, the unaligned bytes count is `fs->flash_parameters->write_block_size - (len % fs->flash_parameters->write_block_size)`, good catch !

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -418,10 +457,18 @@ static int nvs_flash_wrt_entry(struct nvs_fs *fs, uint16_t id, const void *data,
+#ifdef CONFIG_NVS_DATA_CRC
+	/* No CRC has been added if this is a deletion write request */
+	if (len > 0) {
+		entry.len += NVS_DATA_CRC_SIZE;
```
we can move this up to avoid computing crc8 twice when CONFIG_NVS_DATA_CRC is enabled
you can do :
#ifdef CONFIG_NVS_DATA_CRC
if (len > 0) {
		entry.len = (uint16_t)len + NVS_DATA_CRC_SIZE;
}
else {
               entry.len = (uint16_t)len;
}

RICCIARDI-Adrien replied:
Yes, it's better to compute the CRC only once.

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -418,10 +457,18 @@ static int nvs_flash_wrt_entry(struct nvs_fs *fs, uint16_t id, const void *data,
 	nvs_ate_crc8_update(&entry);
 
-	rc = nvs_flash_data_wrt(fs, data, len);
+	rc = nvs_flash_data_wrt(fs, data, len, true);
```
This should be true only when CONFIG_NVS_DATA_CRC is enabled right ?

RICCIARDI-Adrien replied:
Currently the CRC computation is disabled when the `CONFIG_NVS_DATA_CRC` is disabled directly in the `nvs_flash_data_wrt()` function, to avoid putting `ifdef` on each `nvs_flash_data_wrt()` call (see https://github.com/zephyrproject-rtos/zephyr/pull/73436/files#diff-d253a798f77267d97d9d96ebf25162efa0338bba3ed8e3a261439477a57444b7R116). Is it OK for you ?

rghaddab replied:
@RICCIARDI-Adrien you can define something like this to avoid ifdefs
#define data_crc_enabled IS_ENABLED(CONFIG_NVS_DATA_CRC)
Then use it as argument

RICCIARDI-Adrien replied:
The code has been reworked based on this suggestion to avoid several `ifdef`.

RICCIARDI-Adrien commented on subsys/fs/nvs/nvs.c:
```diff
@@ -100,25 +100,32 @@ static inline size_t nvs_al_size(struct nvs_fs *fs, size_t len)
 /* basic aligned flash write to nvs address */
 static int nvs_flash_al_wrt(struct nvs_fs *fs, uint32_t addr, const void *data,
-			     size_t len)
+			    size_t len, uint32_t data_crc, bool is_crc_present)
```
You're right, I will move the CRC computation into the function.

Laczen replied:
> You're right, I will move the CRC computation into the function.

IMHO you will need to change the `nvs_flash_data_wrt()` with local crc calculation, call `nvs_flash_al_wrt()` on the aligned part of data, creating a buffer with remainder of data and crc and call `nvs_flash_al_wrt()` on this created buffer.

RICCIARDI-Adrien replied:
I already made a change, I will update the PR to show you how it's working and if it's fine for you.

Laczen replied:
OK, but the naming of the routine should be respected, `nvs_flash_al_wrt()` is used for an aligned write and also used in writing ate's.

Laczen replied:
I'm not convinced by the approach, it would be better to keep the adding of the crc in the `nvs_flash_data_wrt()` routine.

RICCIARDI-Adrien replied:
OK, I will rework it. Indeed, this will more logical from the architecture point of view.

RICCIARDI-Adrien commented:
A new version of the PR that should address all comments but maybe the last.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -175,12 +175,51 @@ static int nvs_flash_ate_wrt(struct nvs_fs *fs, const struct nvs_ate *entry)
 
 /* data write */
-static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len)
+static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len, bool compute_crc)
```
A little different approach (hope I didn't overlook anything):
```
static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len, bool compute_crc)
{
	int rc;

	/* Only add the CRC if required */
	if (IS_ENABLED(CONFIG_NVS_DATA_CRC) && (compute_crc) && (len != 0U)) {
		const size_t sa_len = len & ~(fs->flash_parameters->write_block_size - 1);
		const size_t rem_len = len - sa_len;
		const uint32_t data_crc = crc32_ieee(data, len);
		uint8_t buf[rem_len + NVS_DATA_CRC_SIZE];
		uint8_t *data8 = (uint8_t *)data;
		
		rc = nvs_flash_al_wrt(fs, fs->data_wra, data8, sa_len);
		fs->data_wra += sa_len;
		if (rc != 0) {
			goto end;
		}


		memcpy(buf, data8 + sa_len, rem_len);
		memcpy(&buf[rem_len], &data_crc, NVS_DATA_CRC_SIZE);
		rc = nvs_flash_al_wrt(fs, fs->data_wra, buf, sizeof(buf));
		fs->data_wra += nvs_al_size(fs, sizeof(buf));
		if (rc != 0) {
			goto end;
		}
	} else {
		rc = nvs_flash_al_wrt(fs, fs->data_wra, data, len);
		fs->data_wra += nvs_al_size(fs, len);
	}

end:
	return rc;
}
```

RICCIARDI-Adrien replied:
Is there an issue with my current implementation ?

Laczen replied:
> Is there an issue with my current implementation ?

Is is not that there is an issue with the current implementation. I was adding a remark about how to avoid any `#ifdef's` in the code and how to simplify the calculation of `aligned_len` and some other nits that I tought it would be simpler to propose an "alternative" that you could pick idea's from (BTW I think the last `if (rc != 0)` can also be removed).

RICCIARDI-Adrien replied:
OK, thanks ! I will update the PR with some of your proposals.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -175,11 +175,44 @@ static int nvs_flash_ate_wrt(struct nvs_fs *fs, const struct nvs_ate *entry)
+		/* Write as much aligned data as possible, so the CRC can be concatenated at
+		 * the end of the unaligned data later
+		 */
+		aligned_len = len - (len % fs->flash_parameters->write_block_size);
```
This can be simplified to: `aligned_len = len & ~(fs->flash_parameters->write_block_size - 1U);`

RICCIARDI-Adrien replied:
Done !

Laczen commented on doc/services/storage/nvs/nvs.rst:
```diff
@@ -19,11 +19,14 @@ combination of these.
+completed. The actual data of the element can be protected by a different (and optional)
+CRC-32. Use the :kconfig:option:`CONFIG_NVS_DATA_CRC` configuration item to enable the
+the data part CRC.
+**Note:** the data CRC is checked only when the whole data of the element is read.
```
Nit: an extra note could be added to stating that enabling the config option on a nvs system that does not have the crc-32 enabled will make all existing data invalid.

RICCIARDI-Adrien replied:
I added the note !

Laczen reviewed (dismissed):
Apart from the nit regarding enabling on existing nvs systems without crc32 this looks good, thanks.

RICCIARDI-Adrien commented:
> Apart from the nit regarding enabling on existing nvs systems without crc32 this looks good, thanks.

I added this note !

Laczen reviewed (dismissed):
Thanks

rghaddab reviewed (commented):
Changes are Ok for me apart from the small style mistake

RICCIARDI-Adrien commented on doc/services/storage/nvs/nvs.rst:
```diff
@@ -19,11 +19,16 @@ combination of these.
-corrective actions when the data CRC does not match its expected value.
+completed. The actual data of the element can be protected by a different (and optional)
+CRC-32. Use the :kconfig:option:`CONFIG_NVS_DATA_CRC` configuration item to enable the
+the data part CRC.
```
Fixed, thanks !
